File: nocomment.py
#!/usr/bin/env python3

#    Copyright (C) 2020 Captain Foxtrot.

#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Constants
MEMORYSIZE:int = 10000
STACKSIZE:int = 10000

# Initialize code area and code pointer
code:str = ""
codePtr:int = 0

# Initialize memory
memory:bytearray = bytearray(MEMORYSIZE)
memoryPtr:int = 0

# Initialize stack
stack:bytearray = bytearray(STACKSIZE)

# Execute a single instruction (character)
def exec(char: str):

    global memory, memoryPtr, MEMORYSIZE, stack

    if char == "i":
        # Increment
        try: memory[memoryPtr] += 1
        # Integer overflow
        except ValueError: memory[memoryPtr] = 0

    elif char == "d":
        # Decrement
        try: memory[memoryPtr] -= 1
        # Integer underflow
        except ValueError: memory[memoryPtr] = 256

    elif char == "l":
        # Decrement pointer
        if memoryPtr: memoryPtr -= 1
        # About to underflow?
        else: memoryPtr = MEMORYSIZE - 1

    elif char == "r":
        # Increment pointer
        if memoryPtr == 256: memoryPtr = 0
        # About to overflow?
        else: memoryPtr += 1

    elif char == "n":
        # Append value of pointer to stack
        stack.append(memory[memoryPtr])

    elif char == "f":
        # Overwrite value of pointer with top item on stack
        memory[memoryPtr] = stack[-1]
        # Drop an item from the stack
        del stack[-1]

    elif char == "s":
        # Check if the value of the pointer is greater than zero
        if memory[memoryPtr] != 0:
            # Jump forward x number of spaces where x is the value on the top of the stack
            codePtr += stack[-1]

            assert codePtr < len(code)
        # Drop an item from the stack
        del stack[-1]

    elif char == "b":
        # Check if the value of the pointer is greater than zero
        if memory[memoryPtr] != 0:
            # Jump forward x number of spaces where x is the value on the top of the stack
            codePtr -= stack[-1]

            assert codePtr >= 0
        # Drop an item from the stack
        del stack[-1]

    elif char == "o":
        # Print the value of the pointer as an ASCII character, without a newline
        print(chr(memory[memoryPtr]), end = "")

    # Any characters that are not commands will throw an error.
    else:
        raise Exception("Unrecognized command: " + char)

# Pass a string of NoComment code into this function to run it.
def exec_str(_code: str):

    global codePtr

    while codePtr < len(_code):
        exec(_code[codePtr])
        codePtr += 1
        
# This part will not be run if this script is being run through an import.
if __name__ == "__main__":

    import sys

    # Read the contents of the file into `code`
    try:
        FILENAME:str = sys.argv[1]
        with open(FILENAME, "r") as codeFile:
            code = codeFile.read()

    # User did not enter a file to run, enter REPL
    except IndexError:
        print("Enter the name of a file to run.")

    # File does not exist
    except FileNotFoundError:
        print(f"File {FILENAME} does not exist.")
        sys.exit(1)

    # Execute the code
    exec_str(code)

File: nocomment.c
#include        <stdio.h>
#include        <stdint.h>
#include        <stdlib.h>

#define         MEMORYSIZE      10000
#define         STACKSIZE       10000

FILE*           f;
long            filelen;
uint8_t*        memory;
uint8_t*        memptr;
uint8_t*        stackbase;
uint8_t*        stackptr;
char            ch;

void free_allocated(void){
    fclose(f);
    free(memory);
    free(stackbase);
}

int main(const int argc, const char** argv){

        // Exit if no filename provided
        if(argc == 1){
                printf("Enter a filename to run.\n");
                return 1;
        }

        // Exit if the file could not be opened
        if(!(f = fopen(argv[1], "r"))){
                printf("Unable to open file \"%s\".\n", argv[1]);
                return 1;
        }

        // Allocate memory for the memory space and stack
        memory    = memptr   = (uint8_t*) calloc(MEMORYSIZE, sizeof(uint8_t));
        stackbase = stackptr = (uint8_t*) calloc(STACKSIZE,  sizeof(uint8_t));

        // Print error message and exit if memory space allocation failed
        if(!memory)
                goto allocationerror;

        // Print error message and exit if stack allocation failed
        if(!stackbase)
                goto allocationerror;
        
        // https://stackoverflow.com/questions/238603/how-can-i-get-a-files-size-in-c
        fseek(f, 0, SEEK_END);
        filelen = ftell(f);
        rewind(f);

        while(1){
                // Get the next character in the file
                ch = fgetc(f);

                // Exit loop if end of file
                if(feof(f))
                        break;
                
                // 
                switch(ch){

                        // Increment value of pointer
                        case 'i':
                                ++*memptr;
                                break;

                        // Decrement value of pointer
                        case 'd':
                                --*memptr;
                                break;

                        // Clear (zero) value of pointer
                        case 'c':
                                *memptr = 0;
                                break;

                        // Decrement pointer
                        case 'l':
                                --memptr;

                                // Wrap-around if pointer underflow
                                if(memptr < memory)
                                        memptr += MEMORYSIZE;
                                break;

                        // Increment pointer
                        case 'r':
                                ++memptr;

                                // Wrap-around if pointer overflow
                                if((int)memory + MEMORYSIZE == (int)memptr)
                                        memptr = memory;
                                break;

                        // Push value of pointer to stack
                        case 'n':
                                // Check for stack overflow
                                if(stackptr >= stackbase + STACKSIZE){
                                        printf("Stack overflow.\n");
                                        free_allocated();
                                        return 1;
                                }

                                // Put the ptr val on stack
                                *stackptr = *memptr;

                                // Increment stack pointer
                                ++stackptr;

                                break;

                        // Pop from stack to value of pointer
                        case 'f':
                                // Check for stack underflow
                                if(stackptr <= stackbase){
                                        printf("Stack underflow.\n");
                                        free_allocated();
                                        return 1;
                                }

                                // Decrement stack pointer
                                --stackptr;

                                // Override ptr val with stack
                                *memptr = *stackptr;

                                break;

                        // Forward jump if non-zero value of pointer
                        case 's':
                                if(*memptr){
                                        // If the jump will not take us out of bounds
                                        if(ftell(f) + *(stackptr - 1) <= filelen)
                                                // Perform jump
                                                fseek(f, *(stackptr - 1), SEEK_CUR);
                                        else {
                                                printf("Code pointer out of bounds.\n");
                                                free_allocated();
                                                return 1;
                                        }
                                }
                                break;

                        // Backwards jump if non-zero value of pointer
                        case 'b':
                                if(*memptr){
                                        // If the jump will not take us out of bounds
                                        if(*(stackptr - 1) - 1 <= ftell(f))
                                                // Perform jump
                                                fseek(f, -*(stackptr - 1), SEEK_CUR);
                                        else {
                                                // Print error message, exit
                                                printf("Code pointer out of bounds.\n");
                                                free_allocated();
                                                return 1;
                                        }
                                }
                                break;

                        // Print value of pointer as ASCII char
                        case 'o':
                                putchar(*memptr);
                                break;

                        // Unknown instruction
                        default:
                                printf("Illegal instruction %i at character %li.\n", ch, ftell(f));

                                free_allocated();
                                return 1;
                }
        }

        // Close the file and deallocate memory
        free_allocated();

        // Exit
        return 0;

        allocationerror:
                printf("Unable to allocate memory.\n");
                free_allocated();
                return 1;
}
